FirebaseAuth/Sources/Utilities/FIRAuthRecaptchaVerifier.m

```diff
@@ -0,0 +1,252 @@
+/*
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#import "FirebaseAuth/Sources/Utilities/FIRAuthRecaptchaVerifier.h"
+
+#import "FirebaseAuth/Sources/Auth/FIRAuth_Internal.h"
+#import "FirebaseAuth/Sources/Backend/FIRAuthBackend.h"
+#import "FirebaseAuth/Sources/Backend/RPC/FIRGetRecaptchaConfigRequest.h"
+#import "FirebaseAuth/Sources/Backend/RPC/FIRGetRecaptchaConfigResponse.h"
+#import "FirebaseAuth/Sources/Backend/RPC/FIRVerifyPasswordRequest.h"
+#import "FirebaseAuth/Sources/Public/FirebaseAuth/FIRAuth.h"
+#import "FirebaseAuth/Sources/Utilities/FIRAuthErrorUtils.h"
+
+#if TARGET_OS_IOS
+#import <RecaptchaInterop/RCAActionProtocol.h>
+#import <RecaptchaInterop/RCARecaptchaProtocol.h>
+#endif
+
+static const NSDictionary *providerToStringMap;
+static const NSDictionary *actionToStringMap;
+
+static NSString *const kClientType = @"CLIENT_TYPE_IOS";
+static NSString *const kRecaptchaVersion = @"RECAPTCHA_ENTERPRISE";
+static NSString *const kFakeToken = @"NO_RECAPTCHA";
+
+@implementation FIRAuthRecaptchaConfig
+
+@end
+
+@implementation FIRAuthRecaptchaVerifier
+
++ (id)sharedRecaptchaVerifier:(nullable FIRAuth *)auth {
+  static FIRAuthRecaptchaVerifier *sharedRecaptchaVerifier = nil;
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    sharedRecaptchaVerifier = [[self alloc] init];
+
+    providerToStringMap = @{@(FIRAuthRecaptchaProviderPassword) : @"EMAIL_PASSWORD_PROVIDER"};
+
+    actionToStringMap = @{
+      @(FIRAuthRecaptchaActionSignInWithPassword) : @"signInWithPassword",
+      @(FIRAuthRecaptchaActionGetOobCode) : @"getOobCode",
+      @(FIRAuthRecaptchaActionSignUpPassword) : @"signUpPassword"
+    };
+  });
+  if (sharedRecaptchaVerifier.auth != auth) {
+    sharedRecaptchaVerifier.agentConfig = nil;
+    sharedRecaptchaVerifier.tenantConfigs = nil;
+    sharedRecaptchaVerifier.auth = auth;
+  }
+  return sharedRecaptchaVerifier;
+}
+
+- (NSString *)siteKey {
+  if ([FIRAuth auth].tenantID == nil) {
+    return self->_agentConfig.siteKey;
+  } else {
+    FIRAuthRecaptchaConfig *config = self->_tenantConfigs[[FIRAuth auth].tenantID];
```

Comment:
This is hardcoded to the default app. Please update to use the right instance.

Should it be sharedRecaptchaVerifier.auth?

Reply:
That's right, thanks for catching this!